src/Pos/Generator/BlockEvent.hs

```diff
@@ -41,13 +91,334 @@ data BlockRollbackResult
+genBlockEvents ::
+       (MonadBlockGen ctx m, RandomGen g)
+    => BlockEventGenParams
+    -> RandT g m [BlockEvent]
+genBlockEvents begp = do
+    preBlockEvents <- genBlockEvents' begp
+    let
+        (blockIndexStart, blockIndexEnd) =
+            getBlockIndexRange preBlockEvents
+        blockCount = BlockCount $
+            fromIntegral (blockIndexEnd - blockIndexStart)
+    blocks <- lift $ genBlocks $ BlockGenParams
+        { _bgpSecrets    = begp ^. begpSecrets
+        , _bgpBlockCount = blockCount
+        }
+    let
+        toZeroBased :: BlockIndex -> Int
+        toZeroBased i =
+            getBlockIndex i - getBlockIndex blockIndexStart
+        getBlock :: BlockIndex -> BlundDefault
+        getBlock i =
+            -- Indexing is safe here because by construction we have enough
+            -- blocks for the lookup to succeed.
+            getOldestFirst blocks !! toZeroBased i
+    return $ fmap getBlock <$> preBlockEvents
+
+data GenBlockEventState
+    = GbesStartingState
+    | GbesBlockchain (OldestFirst NE BlockIndex)
+    | GbesExpectedFailure
+
+genBlockEvents' ::
+       (Monad m, RandomGen g)
+    => BlockEventGenParams
+    -> RandT g m [BlockEvent' BlockIndex]
+genBlockEvents' begp =
+    flip evalStateT GbesStartingState $ do
+        events <- replicateWhileM
+            (fromIntegral $ begp ^. begpBlockEventCount)
+            (genBlockEvent begp)
+        finalEvent <- get <&> \case
+            GbesStartingState -> []
+            GbesExpectedFailure -> []
+            GbesBlockchain blockchain ->
+                -- Rollback the entire blockchain.
+                [BlkEvRollback $ BlockEventRollback
+                    { _berInput    = toNewestFirst blockchain
+                    , _berOutValid = BlockRollbackSuccess
+                    }]
+        return $ events ++ finalEvent
+
+replicateWhileM :: Monad m => Int -> m (Maybe a) -> m [a]
+replicateWhileM n m = go n
+  where
+    go k | k < 1 = return []
+         | otherwise =
+             m >>= \case
+                 Nothing -> return []
+                 Just a  -> (a:) <$> go (k-1)
+
+data RollbackFailureType
+    = RftRollbackExcess
+    | RftRollbackDrop
+    deriving (Enum, Bounded)
+
+instance Random RollbackFailureType where
+    random = randomR (minBound, maxBound)
+    randomR (lo,hi) = runRand $ uniform [lo..hi]
+
+data ApplyFailureType
+    = AftApplyBad
+    | AftApplyNonCont
+    deriving (Enum, Bounded)
+
+instance Random ApplyFailureType where
+    random = randomR (minBound, maxBound)
+    randomR (lo,hi) = runRand $ uniform [lo..hi]
+
+genBlockEvent ::
+       (Monad m, RandomGen g)
+    => BlockEventGenParams
+    -> StateT GenBlockEventState (RandT g m) (Maybe (BlockEvent' BlockIndex))
+genBlockEvent begp = do
+    gbes <- get
+    failure <- lift $ byChance (begp ^. begpFailureChance)
+    when failure $ put GbesExpectedFailure
+    rollback <- lift $ byChance (begp ^. begpRollbackChance)
+    case gbes of
```

Comment:
I _think_ no, but I am not sure too. If this code is valid, can you add a comment about it (because it indeed looks suspicious).